<div style="border:solid Chocolate 2px; padding: 40px">
 
<b> Владислав, привет!👋</b> 

Меня зовут Кирилл Васильев, я буду ревьюером твоего проекта. Я предлагаю общаться на «ты», но если привычнее на «вы», просто скажи об этом! 

Я буду оставлять комментарии в твоем проекте. Пожалуйста, не удаляй их, они могут быть полезны в случае повторной проверки.

Свои комментарии я буду обозначать <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:
  
<br/> 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным. Здесь же я могу давать советы и предложения.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div> 


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>
 
     
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>

<br/>
<div class="alert alert-info">
Рекомендую для твоих комментариев использовать синий или любой другой цветной фон. Я постараюсь ничего не пропустить, но если комментарии будут выделены, то это мне очень поможет!
</div>
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой руководитель будет поступать также. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку.



# Проект спринта 7 "Секреты Темнолесья"

 ***Автор проекта: Черных Владислав Тимурович***
 
 ***Дата сдачи проекта: 18.02.2025 г.***


## Введение
### Цели и задачи проекта

**Целью работы** является подготовка данных об играх, выпущенных за период с 2000 по 2013 год включительно, к анализу о продажах.
    
Для этого необходимо будет познакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных. Перед анализом необходимо отобрать данные по времени выхода игры и категоризовать игры по оценкам пользователей и экспертов и затем выделить топ-7 платформ по количеству игр. В работе предлагается выделить три категории:

- `высокая оценка` — с оценкой от 8 до 10 и от 80 до 100, включая правые границы интервалов.
- `средняя оценка` — с оценкой от 3 до 8 и от 30 до 80, не включая правые границы интервалов.
- `низкая оценка` — с оценкой от 0 до 3 и от 0 до 30, не включая правые границы интервалов.

### Описание данных

Данные `/datasets/new_games.csv` содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board).

### Содержимое проекта

1. Введение
2. Загрузка и знакомство с данными
3. Проверка ошибок в данных и их предобработка
4. Выполнение поставленных задач
5. Итоговый вывод

   
<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 Отлично по описанию. Цели и задачи ясны. Контекст проекта будет понятен любому потенциальному читателю.
 
  
    

## Загрузка и знакомство с данными

In [2]:
# Импортируем библиотеку pandas
import pandas as pd
# Выгружаем данные из датасета new_games.csv в датафрейм df
df=pd.read_csv('/datasets/new_games.csv')

In [3]:
# Выводим информацию о датафрейме
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
# Выведем первые пять строк из датафрейма
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


Датасет состоит из 11 столбцов и 16956 строк. 
Изучим типы данных и их корректность:
- **Строковые данные (object).** Семь столбцов имеют тип данных `object`:
    - Столбцы `Name`, `Platform`, `Genre` содержат строковую информацию (название игры, платформы и жанра), что логично для текстовых данных. Здесь тип данных *object* подходит.
    - Столбцы `EU sales`, `JP sales`, `User Score` содержат информацию о продажах и рейтинге от пользователей (нецелочисленном). Для данных столбцов подойдет тип *float*.
    - Столбец `Rating` хранит рейтинг организации ESRB (всего 7 категорий). В этом случае можно использовать тип *category*, чтобы улучшить производительность и оптимизировать память, если набор значений ограничен и заведомо известен. Но так как указанный столбец нам для анализа не понадобится, оставим его с прежним типом.
- **Числовые значения с плавающей запятой (float64)** Четыре столбца имеют тип данных `float64`:
    - Столбцы `NA sales`, `Other sales` содержит информацию о продажах. Здесь тип данных *float64* подходит. 
    - Столбец `Critic Score` и `Year of Release` содержит информацию о рейтинге от критиков и годе выпуска игры. Они имеют тип *float* из-за того, что имеются пропуски в данных. Изменение типа данных не обязательно, оставим тип *float*.

Большое количество пропусков (порядка половины от всего количества) характерно для столбцов `Critic Score`, `User Score` и `Rating`, которые связаны с рейтингом игр от критиков, пользователей и организации ESRB. 

<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 Здесь все верно по выявленным проблемам в данных. Можно также обратить внимание на несоответствия стилю названий колонок.
 
   

## Проверка ошибок в данных и их предобработка

### Изменение типа написания столбцов датафрейма

In [5]:
# Выведем на экран названия всех столбцов датафрейма
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [6]:
# Приведем все столбцы к стилю snake case
df=df.rename(columns={
    'Year of Release':'Year_of_Release',
    'NA sales': 'NA_sales',
    'EU sales': 'EU_sales',
    'JP sales': 'JP_sales',
    'Other sales': 'Other_sales',
    'Critic Score': 'Critic_Score',
    'User Score': 'User_Score'})
# Выведем на экран Обовленные названия всех столбцов датафрейма
df.columns

Index(['Name', 'Platform', 'Year_of_Release', 'Genre', 'NA_sales', 'EU_sales',
       'JP_sales', 'Other_sales', 'Critic_Score', 'User_Score', 'Rating'],
      dtype='object')

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично по переименованию.  
    
Для приведения к змеиному стилю, чтобы не делать вручную, можно также использовать такой вариант:


`games.columns = games.columns.str.lower().str.replace(' ','_')`
    


</div>

### Изменение типа данных

Как говорилось ранее, столбцы `EU sales`, `JP sales`, `User Score` необходимо преобразовать в тип **float**.
Так как они имеют тип **object**, значит в данных есть строковые значения. Определим их.

In [7]:
# Используя сортировку, найдем текстовые значения в каждом столбце
display(df['EU_sales'].sort_values().tail(7))
display(df['JP_sales'].sort_values().tail(5))
df['User_Score'].sort_values().tail(5)

19          9.2
802     unknown
1394    unknown
446     unknown
1131    unknown
1132    unknown
1612    unknown
Name: EU_sales, dtype: object

12          7.2
1379    unknown
467     unknown
819     unknown
4732    unknown
Name: JP_sales, dtype: object

16951    NaN
16952    NaN
16953    NaN
16954    NaN
16955    NaN
Name: User_Score, dtype: object

В столбцах `EU sales` и `JP sales` пропуски заполнены значением "unknown". Заменим его на значение NaN, чтобы в дальнейшем подсчитать пропуски.

In [8]:
# Изменим значения "unknown" на пропуски
import numpy as np
df['EU_sales']=df['EU_sales'].replace('unknown', np.nan)
df['JP_sales']=df['JP_sales'].replace('unknown', np.nan)
# Преобразуем в тип float
df['EU_sales']=pd.to_numeric(df['EU_sales'])
df['JP_sales']=pd.to_numeric(df['JP_sales'])
#Выведем результат
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year_of_Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA_sales         16956 non-null  float64
 5   EU_sales         16950 non-null  float64
 6   JP_sales         16952 non-null  float64
 7   Other_sales      16956 non-null  float64
 8   Critic_Score     8242 non-null   float64
 9   User_Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB



<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 
    
Можно вспомнить из тренажера такой вариант: `df['признак'] = pd.to_numeric(df['признак'], errors='coerce')`
</div>   

Таким образом, в столбце `EU_sales` шесть пропусков, а в столбце `JP_sales` - четыре. 
Далее, займемся столбцом `User_Score`.

На первый взгляд показалось, что в столбце `User_Score` нет строковых значений. Однако в ходе преобразования типа данных на *float* программа выдавала ошибку `Unable to parse string "tbd" at position 302`. Значит в столбце содержится строковое значение `tbd`, что обозначает "to be determined" (в процессе определения). Проверим сколько таких значений.

In [9]:
# Проверяем количество строк со значением "tbd" в колонке User_Score
df[df['User_Score']=='tbd']

,Name,Platform,Year_of_Release,Genre,NA_sales,EU_sales,JP_sales,Other_sales,Critic_Score,User_Score,Rating
119,Zumba Fitness,Wii,2010.0,Sports,3.45,2.59,0.0,0.66,NaN,tbd,E
302,Namco Museum: 50th Anniversary,PS2,2005.0,Misc,2.08,1.35,0.0,0.54,61.0,tbd,E10+
522,Zumba Fitness 2,Wii,2011.0,Sports,1.51,1.03,0.0,0.27,NaN,tbd,T
647,uDraw Studio,Wii,2010.0,Misc,1.65,0.57,0.0,0.20,71.0,tbd,E
659,Frogger's Adventures: Temple of the Frog,GBA,NaN,Adventure,2.15,0.18,0.0,0.07,73.0,tbd,E
...,...,...,...,...,...,...,...,...,...,...,...
16935,Planet Monsters,GBA,2001.0,Action,0.01,0.00,0.0,0.00,67.0,tbd,E
16937,Bust-A-Move 3000,GC,2003.0,Puzzle,0.01,0.00,0.0,0.00,53.0,tbd,E
16938,Mega Brain Boost,DS,2008.0,Puzzle,0.01,0.00,0.0,0.00,48.0,tbd,E
16945,Plushees,DS,2008.0,Simulation,0.01,0.00,0.0,0.00,NaN,tbd,E


Значений `tbd` оказалось 2464 штук. Заменим их на пропуски *None*.

In [10]:
# Изменим значения "tbd" на пропуски и преобразуем в тип float через метод to_numeric() 
# и параметр errors='coerce' (преобразует нечисловое значение в пропуск) в одно действие:
df['User_Score']=pd.to_numeric(df['User_Score'], errors='coerce')
#Выведем результат
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year_of_Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA_sales         16956 non-null  float64
 5   EU_sales         16950 non-null  float64
 6   JP_sales         16952 non-null  float64
 7   Other_sales      16956 non-null  float64
 8   Critic_Score     8242 non-null   float64
 9   User_Score       7688 non-null   float64
 10  Rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB



<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Ага, отлично )
</div>   

Таким образом, мы заменили столбцы `EU sales`, `JP sales`, `User Score` с типа object на тип float двумя разными способами: через метод **.replace()** и через метод **to_numeric() с параметром errors='coerce'**.
В результате в столбцах `EU sales`, `JP sales`, `User Score` появились пропуски, которые будем анализировать далее.

### Поиск и обработка пропущенных значений

В этом разделе посмотрим на количество пропущенных значений по столбцам датафрейма.

In [11]:
# Выводим количество пропущенных строк в датафрейме
df.isna().sum().sort_values(ascending=False)

User_Score         9268
Critic_Score       8714
Rating             6871
Year_of_Release     275
EU_sales              6
JP_sales              4
Name                  2
Genre                 2
Platform              0
NA_sales              0
Other_sales           0
dtype: int64

In [12]:
# Подсчитываем процент строк с пропусками
round(df.isna().sum()/len(df)*100, 2).sort_values(ascending=False)

User_Score         54.66
Critic_Score       51.39
Rating             40.52
Year_of_Release     1.62
EU_sales            0.04
JP_sales            0.02
Name                0.01
Genre               0.01
Platform            0.00
NA_sales            0.00
Other_sales         0.00
dtype: float64


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Верно по вычислениям.
</div>   

В данных наблюдаются пропуски в следующих столбцах:
- `Name` и `Genre`: в 2 строках (0.01% данных) отсутствует информация о названии игры. Возможно пропуски возникли из-за технической ошибки. Можно или удалить, или оставить строки с пропущенными именами игры. В данной работе оставим строки.
- `Year_of_Release`: в 275 строках (1.62% данных) отсутствует информация о годе выхода игры. Возможно пропуски возникли из-за технической ошибки, или в отдельных случаях год не заполняли, так как он присутствует в названии игры. Для ускорения выполнения в работе принято решение удалить пропуски (так как их меньше 2%). Альтернативный вариант- можно заполнить указанные  пропуски с помощью сети Интернет.
- `Rating`: в 6871 строках (40.52% данных) отсутствует информация о рейтинге организации ESRB. Возможно информация по ним отсутствовала и их не заполняли. Указанный столбец для анализа не нужен. Заменим пропуски на значение-индикатор..
- `Critic_Score` и `User_Score `: в 8714 (51.39%) и 9268 (54.66%) строках соотвественно информация отсутствует. Эти показатели показывают рейтинг, поставленный критиками и игроками. Возможно информация отсутствует из-за того, что не было оценок по данным играм, либо по технической ошибке. Так как эти данные необходимы для дальнейшего анализа, удалить их нельзя. Чтобы не повлияли на анализ мы не можем их заменить на средние значения, так как пропусков наравне с имеющимися данными. Предлагается в работе пропуски заменить на значение-индиикатор.
- `EU_sales` и `JP_sales`: в 6 (0.04%) и 4 (0.02%) строках соотвественно отсутствует информация. На месте пропусков были значения unknown. Видимо информация по этим данным отсутствовала. В работе предлагается пропуски заменить на среднее значение в зависимости от названия платформы и года выхода игры.


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Гипотезы появления пропусков есть. Для некоторых признаков, вполне возможно, у нас просто мало данных из-за того, что игра вышла давно (тогда не было рейтингов) или в регионе, которым не интересуются в США (рейтинги США не всегда есть у японских игр, например).
</div>   


 <div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
    
По аргументации заполнения пропусков согласен не везде.
 
В случае рейтингов обработка пропусков будет обязательным действием. При этом заполнение средним будет слишком неточным, ведь рейтинги − вещь очень индивидуальная. Исказить данные нам бы не хотелось. Если здесь заполним средним, то рискуем получить искаженные пропорции игр по рейтингам.
    
В нашем случае хорошим вариантом будет заполнить нереалистичным значением−заглушкой типа «-1», по которому затем можно будет легко отфильтровать данные.

***
    
В случае rating (строковый тип данных) можем заполнить строковой заглушкой типа unknown.
    
Безопасное и оптимальное решение для категориального признака.
    



Удалим пропуски в столбце `Year_of_Release`. Для этого используем метод *dropna()*. Назовем новый датафрейм с обработанными пропусками как **df_not_null**.

После удаления пропусков в столбце `Year_of_Release` в итоговом датафрейме осталось 16956-275=**16681** заполненных значений.

In [30]:
# Удалим пропуски в столбце 'Year_of_Release'
df_not_null=df.dropna(subset=['Year_of_Release'])
# покажем количество пропусков в новом датафйреме df_not_null
df_not_null.isna().sum()

Name                  2
Platform              0
Year_of_Release       0
Genre                 2
NA_sales              0
EU_sales              6
JP_sales              4
Other_sales           0
Critic_Score       8596
User_Score         9123
Rating             6780
dtype: int64

<br/>
<div class="alert alert-info">
А вообще стоит ли удалять значения с годами, если можно найти год выпуска игры в Интернете?

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Если можно найти и есть техническая возможность это реализовать, то это прекрасный вариант заполнения пропусков. Тогда да, логично не удалять.
    
Здесь также можно в приницпе оставить как есть (нежелательный, но не критичный вариант), а также заполнить нереалистиным значением (типа -1 или 0).
</div>
  


<br/>
<div class="alert alert-info">
Хорошо, я исправлю!
    А вообще стоит ли делать значение-индикатор, ведь можно и смотреть данные по пропускам? И фильтровать также можно по пропускам.
</div>    

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Здесь есть некоторая логика, что предобработку делаем как-будто бы для команды и хорошо бы сделать так, чтобы даже наши коллеги понимали, что с данными все ок.
    
Вообще, если пропуски гарантированно не доставляют неудобств, то можно оставить как есть. На вычисления не влияют, нам работать с ними удобно − подводных камней для аналитики, кажется, нет. В DS пропуски обычно очень нежелательны, но и «нереалистичное значение» там нам бы тоже не подошло. В аналитике, на мой взгляд, очень удобно фильтровать по «-1». Ну и действительно, как правило люди без объяснений понимают, что это за категория.
    
То есть, выше у тебя не прямо ошибка-ошибка, но все-таки задание проекта предполагает, что мы обкатаем в коде разные подходы к обработке пропусков, поэтому уж совсем все оставлять как есть будет нехорошо.
    
</div>
  


Заменим пропуски в столбцах с рейтингом `Critic_Score` и `User_Score ` на значение-индикатор, к примеру на -1 (отрицательного рейтинга не бывает). Для столбца `Rating` изменим на значение *unknown*.

In [35]:
df_not_null['Critic_Score']=df_not_null['Critic_Score'].fillna(-1)
df_not_null['User_Score']=df_not_null['User_Score'].fillna(-1)
df_not_null['Rating']=df_not_null['Rating'].fillna('unknown')
#Проверяем результат по наличию пропусков
df_not_null.isna().sum()

Name               2
Platform           0
Year_of_Release    0
Genre              2
NA_sales           0
EU_sales           0
JP_sales           0
Other_sales        0
Critic_Score       0
User_Score         0
Rating             0
dtype: int64

Заменим пропуски в столбцах `EU_sales` и `JP_sales` на среднее значение в зависимости от названия платформы и года выхода игры.
Для этого используем метод *apply()*.

In [36]:
# Напишем функцию для замены пропуска на среднее значение в столбце 'EU_sales':
def fill_sale_EU(row):
    if pd.isna(row['EU_sales']):
        group=df_not_null[(row['Platform']==df_not_null['Platform']) & (row['Year_of_Release']==df_not_null['Year_of_Release'])]
        return group['EU_sales'].mean()
    else:
        return row['EU_sales']
# Заменим пропуски в датафрейме df_not_null:
df_not_null['EU_sales']=df_not_null.apply(fill_sale_EU, axis=1)

# Напишем функцию для замены пропуска на среднее значение в столбце 'JP_sales':
def fill_sale_JP(row):
    if pd.isna(row['JP_sales']):
        group=df_not_null[(row['Platform']==df_not_null['Platform']) & (row['Year_of_Release']==df_not_null['Year_of_Release'])]
        return group['JP_sales'].mean()
    else:
        return row['JP_sales']
# Заменим пропуски в датафрейме df_not_null:
df_not_null['JP_sales']=df_not_null.apply(fill_sale_JP, axis=1)
# Посчитаем количество пропусков в обработанном датафрейме:
import warnings
warnings.filterwarnings('ignore')
df_not_null.isna().sum()

Name               2
Platform           0
Year_of_Release    0
Genre              2
NA_sales           0
EU_sales           0
JP_sales           0
Other_sales        0
Critic_Score       0
User_Score         0
Rating             0
dtype: int64

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 По заполнению пропусков здесь удобно также использовать функцию .transform() после .groupby()


```
#Заполнение продаж средним значением
avg_eu_sales = df.groupby(['year_of_release', 'platform'])['eu_sales'].transform('mean')
df['eu_sales'] = df['eu_sales'].fillna(avg_eu_sales)

avg_jp_sales = df.groupby(['year_of_release', 'platform'])['jp_sales'].transform('mean')
df['jp_sales'] = df['jp_sales'].fillna(avg_jp_sales)
```



Подробнее можно почитать <a href="https://dfedorov.spb.ru/pandas/%D0%9F%D0%BE%D0%BD%D0%B8%D0%BC%D0%B0%D0%BD%D0%B8%D0%B5%20%D1%84%D1%83%D0%BD%D0%BA%D1%86%D0%B8%D0%B8%20transform%20%D0%B2%20Pandas.html">здесь</a>



<div class="alert alert-info">
Спасибо! Буду теперь знать, этим способом намного легче решать.

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Посмотри, что мы получаем в промежуточном шаге avg_eu_sales для понимания работы transform. Там будет просто список значений с индексами размера количества строк в таблице. Для каждой строки мы получаем среднее по году и платформе. Во второй строке мы оставляем наши данные как есть в тех строках, где нет пропусков, а где пропуски есть, вставляем соответствующее значение из соответствующей строки списка transform. 
</div>
  


<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

В нашем случае предупреждение говорит о том, что мы ссылаемся на срез данных и при изменениях в срезе изменения коснутся и основного датасета. Если такая ситуация для нас нежелательна, лучше предварительно создать копию датасета и работать с ней. Например, так: df_1 = df.copy(). Если же фактически мы ничего не изменяем, можем игнорировать предупреждение.

В финальной версии проекта рекомендуется после изучения предупреждений и принятия по ним решений убрать их.

Убрать можно так: `warnings.filterwarnings('ignore')` после `import warnings`.

Или так:

```import warnings
warnings.simplefilter(action='ignore')
```



<div class="alert alert-info">
По идее же я меняю данные не в исходном датасете, а в новом df_not_null.
    В любом случае применю указанную твою функцию, чтобы предупреждения не пугали "Заказчика".

<div class="alert alert-warning">
<h2> Комментарий ревьюера  #2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Эту таблицу мы получаем так: `df_not_null=df.dropna(subset=['Year_of_Release'])`
    
Я зашел в документацию, но там прямо не указывается, как именно «под капотом» работает функция. Я предполагаю (по аналогии с другими подобными функциями и потому, что видим с предупреждением), что здесь мы просто создаем срез из изначального датасета и сохраняем его в новую переменную. Тогда как в датасете df у нас будут по прежнему храниться изначальные данные. Вот если бы мы реализовали `df=df.dropna(subset=['Year_of_Release'])`, то перезаписали бы наш датасет и он был бы уже без дубликатов. Обычно так и делают. Нам дубли не нужны в принципе, поэтому и не нужен вариант датасета с этими дублями. 
    
Срез, сохраненный в переменную, можно представлять себе как ярлык в Виндоус - это просто ссылка на некоторый файл, который лежит в другом месте. В нашем случае ссылка с фильтрацией на изначальную таблицу. При чем «ленивый» интерпретатор python начнет делать фильтрацию только тогда, когда мы запустим какое-то вычисление, в котором эта фильтрация нужна. Такая реализация позволяет сохранять память, не плодить лишних копий датасета, однако может приводить к нежелательным последствиям, если не учитывать эту особенность.
    
Если это так, то предупреждение логично: мы обращаемся к некоторому срезу данных, который ссылается на изначальную таблицу. Если мы будем что-то менять в срезе, то изменения будут применены и к изначальной таблице df (ниже по ссылке есть примеры, как можно посмотреть, что это действительно сработает).
    
Поэтому язык python нам так докучает с этим предупреждением. На деле, правда, представить ситуацию, когда мы случайно изменим изначальные данные, сложно. Нужно прямо сильно постараться, чтобы такое случилось. Поэтому обычно рекомендуют просто игнорировать это предупреждение.
    
Однако чтобы полностью исключить нежелательное поведение кода, если нам это нужно, мы можем явно создать копию среза и сохранить данные в отдельном датасете. Для этого нужно будет воспользоваться copy().
    
Вот здесь можно почитать про копирование и посмотреть на поведение датасета, если обращаться к нему через срез, который н него ссылается: https://pythonworld.ru/moduli/modul-copy.html 
       
</div>




In [42]:
import warnings
warnings.filterwarnings('ignore')

**Таким образом, в ходе работы с пропусками было сделано:**
- удалены 275 строк с пропусками в столбце `Year_of_Release`;
- заменены пропуски в столбцах `EU_sales` и `JP_sales` на среднее значение в разрезе названия платформы и года выпуска игры;
- пропуски в столбцах `Critic_Score`, `User_Score` и `Rating` были заменены на значение индикатор.
- пропуски в столбцах `Name` и `Genre` оставлены. Обработка пропущенных значений будет проводиться в дальнейшем (при необходимости).

 <div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Рейтинги (три признака) обязательно нужно обработать. Два из них дальше будем использовать в категоризации, а для rating есть хорошее безопасное решение. 
    
Есть еще такая логика, что пропуски лучше не оставлять, так как предобработку можем делать не только для себя, но и для коллег. Если пропусков не будет, коллеги будут больше уверены в том, что предобработку сделали и ничего не забыли.

<div class="alert alert-info">
Сделано!

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Спасибо)
</div>
  


### Обработка дубликатов

Посмотрим строки с пропусками в столбцах `Name`.

In [43]:
df_not_null[df_not_null['Name'].isna()]

,Name,Platform,Year_of_Release,Genre,NA_sales,EU_sales,JP_sales,Other_sales,Critic_Score,User_Score,Rating
661,NaN,GEN,1993.0,NaN,1.78,0.53,0.03,0.08,-1.0,-1.0,unknown
14439,NaN,GEN,1993.0,NaN,0.00,0.00,0.03,0.00,-1.0,-1.0,unknown


Как видно, пропуски с названиями игр и жанрами совпадают. Также видно, что для этих строк одинаковая платформа и год выпуска. 
Если посмотреть на столбцы с продажами, то для игры №661 отсутствуют продажи в Японии, а для игры №14439 - отстутсвуют продажи в остальных странах. Можно сделать вывод, что это одна и та же игра, которую заполняли дважды.
Добавим в игру №661 продажи в Японии, а игру №14439 удалим.

<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Наблюдение хорошее, но я бы обе игры удалил без названия. 

<div class="alert alert-info">
С твоего позволения оставлю в данной работе.

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Без проблем, решение рабочее.
</div>
  


In [44]:
# Копируем датафрейм df_not_null в df_not_dupl
df_not_dupl=df_not_null
# Заменим значение для игры под индексом 661 в столбце JP_sales на 0.03 (взято из строки №14439)
df_not_dupl.loc[661, 'JP_sales']=0.03
# Удалим строку с индексом 14439 с помощью метод drop()
df_not_dupl=df_not_dupl.drop(14439)
# Посмотрим результат
df_not_dupl.loc[df_not_dupl['Name'].isna()]

,Name,Platform,Year_of_Release,Genre,NA_sales,EU_sales,JP_sales,Other_sales,Critic_Score,User_Score,Rating
661,NaN,GEN,1993.0,NaN,1.78,0.53,0.03,0.08,-1.0,-1.0,unknown


Таким образом, избавились от одного неявного дубликата. В любом случае, на дальнейший анализ он не повлиял бы, так год выпска игры не попадает в интересующий нас интервал.

Далее, проанализируем  уникальные значения в категориальных данных: жанр игры, платформа, рейтинг и год выпуска.

In [45]:
df_not_dupl['Rating'].value_counts().head(100).sort_index()

AO            1
E          3968
E10+       1414
EC            8
K-A           3
M          1558
RP            1
T          2948
unknown    6779
Name: Rating, dtype: int64

 <div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь хитрая ситуация. У нас есть игры как из Европы, так и из Азии. В основном используется рейтинг ESRB https://ru.wikipedia.org/wiki/Entertainment_Software_Rating_Board 
    
Но некоторых категорий из представленных у нас в нем нет. Фактически, это сходные категории, но в другом обозначении.
    
Например, можно заменить значения 'К-А' на 'Е'.
    
Рекомендую в подобных случаях уточнять информацию в интернете (в реальных проектах на реальной работе также, ну и можно всегда спрашивать у коллег, особенно у тех, кто предоставил данные. Часто такие категории объясняются в паспорте данных. Ну в нашем случае нужно залезть в интернет). 




<div class="alert alert-info">
    На самом деле я заметил, что 'K-A' отсутствует в перечне ESRB. Из-за того, что столбец 'Rating' не нужен был в анализе, я решил его не трогать.
    Сейчас поправим!

Дубликаты в столбце `Rating` отсутствуют. Однако, как можно заметить, рейтинг `K-A` в актуальном перечне ESRB отсутствует. Рейтинг `K-A` это старое название рейтинга `E`. Сделаем замену:

In [48]:
# Заменим 'K-A' на 'Е'
df_not_dupl['Rating']=df['Rating'].replace('K-A', 'E')
# Проверим результат замены
df_not_dupl['Rating'].value_counts().head(100).sort_index()

AO         1
E       3971
E10+    1414
EC         8
M       1558
RP         1
T       2948
Name: Rating, dtype: int64

<div class="alert alert-info">
Замена сделана!

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 
</div>
  


In [49]:
df_not_dupl['Year_of_Release'].value_counts().head(100).sort_index()

1980.0       9
1981.0      46
1982.0      37
1983.0      18
1984.0      14
1985.0      14
1986.0      22
1987.0      17
1988.0      15
1989.0      17
1990.0      16
1991.0      42
1992.0      43
1993.0      61
1994.0     121
1995.0     220
1996.0     267
1997.0     293
1998.0     384
1999.0     341
2000.0     357
2001.0     491
2002.0     839
2003.0     789
2004.0     771
2005.0     950
2006.0    1020
2007.0    1218
2008.0    1445
2009.0    1450
2010.0    1279
2011.0    1149
2012.0     670
2013.0     552
2014.0     584
2015.0     612
2016.0     507
Name: Year_of_Release, dtype: int64

Дубликаты в столбце `Year_of_Release` отсутствуют.

In [50]:
df_not_dupl['Platform'].value_counts().head(100).sort_index()

2600     118
3DO        3
3DS      522
DC        52
DS      2147
GB        97
GBA      826
GC       549
GEN       28
GG         1
N64      320
NES      100
NG        12
PC       972
PCFX       1
PS      1208
PS2     2154
PS3     1330
PS4      395
PSP     1212
PSV      434
SAT      174
SCD        6
SNES     241
TG16       2
WS         6
Wii     1305
WiiU     147
X360    1249
XB       818
XOne     251
Name: Platform, dtype: int64

Дубликаты в столбце `Platform` отсутствуют.

In [51]:
df_not_dupl['Genre'].value_counts().head(100).sort_index()

ACTION            12
ADVENTURE          4
Action          3341
Adventure       1309
FIGHTING           6
Fighting         844
MISC               3
Misc            1740
PLATFORM           3
PUZZLE             2
Platform         891
Puzzle           577
RACING             6
ROLE-PLAYING       6
Racing          1244
Role-Playing    1493
SHOOTER            5
SIMULATION         1
SPORTS             7
STRATEGY           1
Shooter         1314
Simulation       866
Sports          2325
Strategy         679
Name: Genre, dtype: int64

Дубликаты в столбце `Genre` **присутствуют**. Они появились из-за разного регистра написания жанра.
Приведем к одному регистру.

In [52]:
# переведем жанр игры в нижний регистр
df_not_dupl['Genre']=df_not_dupl['Genre'].str.lower()
# проверим результат
df_not_dupl['Genre'].value_counts().head(100).sort_index()

action          3353
adventure       1313
fighting         850
misc            1743
platform         894
puzzle           579
racing          1250
role-playing    1499
shooter         1319
simulation       867
sports          2332
strategy         680
Name: Genre, dtype: int64

Аналогично поступим и с названием игры `Name`:

In [53]:
# создадим копию датафрейма
df_not_dupl0=df_not_dupl
# посчтитаем количество уникальных игр без обработки регистра
Name_out_reg=df_not_dupl0['Name'].nunique()
# переведем название игры в нижний регистр
df_not_dupl['Name']=df_not_dupl['Name'].str.lower()
# посчтитаем количество уникальных игр после обработки регистра
Name_with_reg=df_not_dupl['Name'].nunique()
print(f'''Количество уникальных игр без обработки регистра: {Name_out_reg}
Количество уникальных игр после обработки регистра: {Name_with_reg}''')

Количество уникальных игр без обработки регистра: 11426
Количество уникальных игр после обработки регистра: 11426


Таким образом, повторяющихся названий игр после обработки регистра выявлено не было.
Неявные дубликаты находились только в столбце `Genre`, от которых успешно избавились в данной работе.
Далее, избавимся от явных дубликатов.

In [54]:
# посчитаем количество явных дубликатов
df_not_dupl.duplicated().sum() 

235

In [55]:
# удаляем дубликаты и сохраняем в новый датафрейм
df_cleaned=df_not_dupl.drop_duplicates()
print(f'''Количество игр без обработки данных: {len(df)}
Количество уникальных игр после обработки: {len(df_cleaned)} ({round(len(df_cleaned)/len(df)*100, 1)}%)''')

Количество игр без обработки данных: 16956
Количество уникальных игр после обработки: 16445 (97.0%)


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍  Точно 
</div> 
        


### Выводы по предварительной обработке данных

Были загружены данные `new_games.csv`, которые содержат 11 столбцов и изначально содержали 16956 строк, в которых представлена информация о продажах игр, сделанных в разных жанрах и выпущенных на разных платформах, а также пользовательские и экспертные оценки игр. 
В ходе предобработки данных было сделано:
- Столбцы `EU sales`, `JP sales`, `User Score` были приведены к числовому типу **float**, так как это необходимо для дальнейшего анализа и категоризации игр по оценкам пользователей.
- Пропуски были выявлены в 8 столбцах:
    - 275 строк с пропусками (1.62%) в столбце `Year_of_Release` были удалены;
    - пропуски в столбцах `Name`, `Genre` были оставлены;
    - пропуски в столбцах `Critic_Score` и `User_Score ` были заменены на значение-индикатор **-1**.
    - пропуски в столбце `Rating` были заменены на значение-индикатор **unknown**. К тому же три значения рейтинга `K-A` были актуализированы и исправлены на `E`.
    - пропуски в столбцах `EU_sales` и `JP_sales` были заменены на средние значения в разрезе названеия платформы и года выпуска.
- Удален один неявный дубликат, который в двух строках содержал одну и ту же информацию.
- Были приведены к нижнему регистру значения в столбцах `Name` и `Genre` для обнаружения неявных дубликатов.
- Удалены 235 явных дубликатов, что составляет около 3% от всего количества данных.

Таким образом, после обработки данных было удалено 511 строк (3% от общего количества). В итоге для дальнейшего анализа осталось **16445 строк**. Далее, перейдем к выполнению поставленных задач.


<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично, что не забываешь о промежуточных выводах.
    
Важно будет уточнить по предобработке рейтингов.
</div>  

<div class="alert alert-info">
Исправил вывод!

## Выполнение поставленных задач

### Фильтрация данных

Необходимо отобрать игры, выпущенные в период с 2000 по 2013 год включительно. Сохраните новый срез данных в отдельном датафрейме df_actual.

In [56]:
# проведем фильтрацию и выберем игры, выпущенные с 2000 по 2013 год включительно
df_actual=df_cleaned[(df_cleaned['Year_of_Release']>=2000.0) & (df_cleaned['Year_of_Release']<=2013.0)]
# проверим результат фильтрации
print(f'Количество удаленных строк: {(len(df_cleaned)-len(df_actual))}')
print(f'Количество данных с 2000 по 2013 год: {len(df_actual)}')
df_actual['Year_of_Release'].value_counts().sort_index()

Количество удаленных строк: 3664
Количество данных с 2000 по 2013 год: 12781


2000.0     350
2001.0     482
2002.0     829
2003.0     775
2004.0     762
2005.0     939
2006.0    1006
2007.0    1197
2008.0    1427
2009.0    1426
2010.0    1255
2011.0    1136
2012.0     653
2013.0     544
Name: Year_of_Release, dtype: int64

Таким образом, при фильтрации игр, выпущенных с 2000 по 2013 год включительно, в срезе данных осталось **12781 строк**.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Все верно.
</div> 
        


### Категоризация данных

Проведем категоризацию данных:
- Разделим все игры по **оценкам пользователей** и выделим такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).
- Разделим все игры по **оценкам критиков** и выделим такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).
Для категоризации данных используем метод *apply(categorize)*, где *categorize* - функция для соотнесения значений ячейки к определенной категории.

In [57]:
# Функция для категоризации
def categorize(value):
    if value<=10.0 and value>=8.0:
        return 'высокая оценка'
    elif value<8.0 and value>3.0:
        return 'средняя оценка'
    elif value<3.0 and value>0.0:
        return 'низкая оценка'
    else:
        return 'без оценки'
# Применим функцию к нашему датафрейму и создаздим новые столбцы 'Оценки критиков' и 'Оценки пользователей'
df_actual['Оценки критиков']=(df_actual['Critic_Score']/10).apply(categorize)
df_actual['Оценки пользователей']=df_actual['User_Score'].apply(categorize)
# Расположим столбцы в удобном порядке с помощью метода reindex()
df_actual=df_actual.reindex(columns=['Name', 'Platform', 'Year_of_Release', 'Genre', 'NA_sales', 'EU_sales', 'JP_sales', 'Other_sales', 'Critic_Score', 'Оценки критиков', 'User_Score', 'Оценки пользователей', 'Rating'])
# Отсортируем значения по оценкам критиков и пользователей в порядке убывания
df_actual.sort_values(by=['Critic_Score', 'User_Score'], ascending=False).head(5)

,Name,Platform,Year_of_Release,Genre,NA_sales,EU_sales,JP_sales,Other_sales,Critic_Score,Оценки критиков,User_Score,Оценки пользователей,Rating
51,grand theft auto iv,X360,2008.0,action,6.76,3.07,0.14,1.03,98.0,высокая оценка,7.9,средняя оценка,M
227,tony hawk's pro skater 2,PS,2000.0,sports,3.05,1.41,0.02,0.20,98.0,высокая оценка,7.7,средняя оценка,T
57,grand theft auto iv,PS3,2008.0,action,4.76,3.69,0.44,1.61,98.0,высокая оценка,7.5,средняя оценка,M
519,metroid prime,GC,2002.0,shooter,1.96,0.67,0.10,0.09,97.0,высокая оценка,9.3,высокая оценка,T
97,super mario galaxy 2,Wii,2010.0,platform,3.56,2.35,0.98,0.62,97.0,высокая оценка,9.1,высокая оценка,E


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍  Хорошее решение с `['Critic_Score']/10`
</div>   
 

Таким образом, мы создали два новых столбца 'Оценки критиков' и 'Оценки пользователей', в которых провели категоризацию по оценкам игры от критиков и пользователей соответсвенно за период с 2000 по 2013 год включительно. Для наглядности датафрейм был отсортирован по оценкам по убыванию (сначала по оценкам критиков, потом уже пользователей).

Далее, проведем анализ оценок игр от критиков и пользователей.

In [60]:
# сгруппируем данные по оценкам критиков и скопируем в датафрейм df_grouped_critics
df_grouped_critics=df_actual.groupby(['Оценки критиков'], as_index=False)['Year_of_Release'].count()
# переименуем колонку с подсчетом игр для каждой категории оценки в 'Count_games'
df_grouped_critics=df_grouped_critics.rename(columns={'Year_of_Release':'Count_games'})
# посчитаем процент каждой категории от общего количества
df_grouped_critics['percent_count']=round(df_grouped_critics['Count_games']/df_grouped_critics['Count_games'].sum()*100, 1)
# покажем результат
display(df_grouped_critics.sort_values(['Count_games'], ascending=False))

# сгруппируем данные по оценкам критиков и скопируем в датафрейм df_grouped_users
df_grouped_users=df_actual.groupby(['Оценки пользователей'], as_index=False)['Year_of_Release'].count()
# переименуем колонку с подсчетом игр для каждой категории оценки в 'Count_games'
df_grouped_users=df_grouped_users.rename(columns={'Year_of_Release':'Count_games'})
# посчитаем процент каждой категории от общего количества
df_grouped_users['percent_count']=round(df_grouped_users['Count_games']/df_grouped_users['Count_games'].sum()*100, 1)
# покажем результат
display(df_grouped_users.sort_values(['Count_games'], ascending=False))

,Оценки критиков,Count_games,percent_count
0,без оценки,5624,44.0
3,средняя оценка,5410,42.3
1,высокая оценка,1692,13.2
2,низкая оценка,55,0.4


,Оценки пользователей,Count_games,percent_count
0,без оценки,6316,49.4
3,средняя оценка,4064,31.8
1,высокая оценка,2286,17.9
2,низкая оценка,115,0.9


Как видно, за период с 2000 по 2013 год включительно большая часть игр получила **среднюю** (42.3% и 31.8% от критиков и пользователей соответственно) или **высокую** оценку (17.9% и 13.2%). Доля игр, получившая низкую оценку, от критиков и от пользователей составляет менее 1.0%. Следует учесть, что в выборке находилось около половины игр, по которым не было информации об оценке от критиков и пользователей.

 
<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Согласен. Здесь все замечательно.
</div>  


### Топ-7 платформ по количеству игр

Покажем топ-7 платформ по количеству игр, выпущенных за период с 2000 по 2013 год включительно.

In [59]:
df_actual['Platform'].value_counts().head(7)

PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: Platform, dtype: int64

Как видно, в лидеры по количеству игр, выпущенных за период с 2000 по 2013 год являются платформы **PS2** и **DS**, в которых было выщено 2127 и 2120 игр соответственно.

 
<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 
</div>  


## Итоговый вывод

В ходе работы было сделано:
1. Проведена предобработка данных датасета `new_games.csv`, в ходе которой из 16956 строк было удалено 511 строк (3%), содержащих пропуски в столбце с годом выпуска игры и содержащих явные/неявные дубликаты.
2. Столбцы `EU sales`, `JP sales`, `User Score` приведены из типа **object** к числовому типу **float**. Пропуски в столбцах `Name`, `Genre` оставлены. Пропуски в столбцах `Rating`, `Critic_Score` и `User_Score ` были заменены на значение-индикатор. Пропуски в столбцах `EU_sales` и `JP_sales` заменены на средние значения в разрезе названеия платформы и года выпуска.
3. Проведена фильтрация игр, выпущенных **с 2000 по 2013 год включительно**, в ходе которой было удалено 3664 строк. Таким образом, в срезе данных для дальнейшего анализа осталось **12781** строк.
4. Проведена категоризация игр по оценкам критиков и пользователей, в ходе которой в исходный датасет были добавлены два дополнительных столбца: `Оценки критиков` и `Оценки пользователей`. Большая часть игр получила **среднюю** (42.3% и 31.8% от критиков и пользователей соответственно) или **высокую оценку** (17.9% и 13.2%). Доля игр, получившая низкую оценку, от критиков и от пользователей составляет менее 1.0%
5. Получено топ-7 платформ по количеству игр. Явными лидерами являются платформы **PS2** и **DS**, в которых было выщено 2127 и 2120 игр соответственно.

Датафрейм `df_actual` можно в дальнейшем использовать для анализа продаж для игр, выпущенных с 2000 по 2013 год включительно.

<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Замечательно по содержанию и оформлению. После доработки нужно перепроверить, что общий вывод синхронизирован с решениями в коде.
</div>  
	  



<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту**
    
Спасибо за твой проект! Внимательная работа, практически все решения верные и оптимальные, но кое-что нужно доработать и кое-что можно улучшить, чтобы проект стал совсем замечательным.
    
**Отмечу положительные моменты проекта🚀**
    
* Отличное оформление. Есть описание, комментарии, выводы.
* Хорошее владение кодом, инструментами предобработки. Все сложные моменты выполнены отлично.
* Хорошая аргументация решений. Во всех случаях понятно, почему принимали те или иные решения.
    
**На что стоит обратить внимание🤔**
    
* Необходимо уточнить по стратегии заполнения пропусков. Важно не пропустить все колонки с рейтингами.
* Эти и остальные замечания я также написал по ходу проекта.
    
**Желаю удачи и жду твой проект на повторное ревью!**😉
    
    
</div>


<div class="alert alert-info">
Выводы поправил. Спасибо за ревью!

<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий ко второй проверке**

Спасибо за внимательную доработку и вопросы. Критичных комментариев нет. 
    
Выше я объяснил, откуда берется предупреждение, если мы все вроде бы сохранили в новую таблицу (на самом деле не сохранили). Также немного проговорил про функцию transform.
    
Если что-то пропустил или остались вопросы, напиши в телеграм @kriill

Желаю удачи в дальнейшей учебе!

</div>
